rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -1059,26 +1055,41 @@ def training_states_labels_and_entities(
 
                 pbar.set_postfix({f"# {self.LABEL_NAME}": f"{len(example_labels):d}"})
 
-        self._remove_user_text_if_intent(example_states)
-
-        # Collect all positive intent labels for a given state hash and add
-        # the set back to the singleton labels.
-        for positive_label_list in state_hash_to_label_list_instances.values():
-            # Get the set of positive labels associated with each state hash.
-            positive_label_set = set([labels[0] for labels in positive_label_list])
-
-            for labels in positive_label_list:
-                # Extend the singletone `labels` with the `positive_label_set`.
-                # Note that we need to filter out the redundant label `labels[0]`
-                # from `positive_label_set` so as not to add it twice.
-                filtered_label_set = filter(
-                    lambda label: label != labels[0], positive_label_set
+        pruned_example_states = []
+        pruned_example_labels = []
+        pruned_example_entities = []
+
+        # All data points with duplicate input states should be pruned
+        # out if `self.remove_duplicates=True` and there should be only
+        # one data point where all positive labels are available for
+        # the corresponding input state.
+        seen_states = set()
+        for example_state, example_entity in zip(example_states, example_entities):
+            state_hash = self._hash_example(example_state)
+            if self.remove_duplicates and state_hash in seen_states:
+                continue
+            pruned_example_states.append(example_state)
+            pruned_example_labels.append(
+                list(
+                    set(
+                        [
+                            labels[0]
```

Comment:
Can you please add a comment explaining why `[0]`?

Reply:
Done